File: examples/go/streaming/client-disconnect/agent.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net"
	"net/http"
	"os"
	"os/signal"
	"time"
)

const (
	kubemqURL = "http://localhost:9090"
	agentID   = "slow-stream-agent-01"
	agentPort = 18080
)

func handleStream(w http.ResponseWriter, r *http.Request) {
	flusher, ok := w.(http.Flusher)
	if !ok {
		http.Error(w, "streaming not supported", http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "text/event-stream")
	w.Header().Set("Cache-Control", "no-cache")
	w.Header().Set("Connection", "keep-alive")

	for i := 1; i <= 10; i++ {
		event, _ := json.Marshal(map[string]interface{}{
			"type": "status_update",
			"payload": map[string]interface{}{
				"status": "working", "progress": i, "total": 10,
			},
		})
		_, err := fmt.Fprintf(w, "event: task.status\ndata: %s\n\n", event)
		if err != nil {
			fmt.Printf("Client disconnected after event %d\n", i-1)
			return
		}
		flusher.Flush()
		fmt.Printf("Sent event %d\n", i)
		time.Sleep(2 * time.Second)
	}

	done, _ := json.Marshal(map[string]interface{}{
		"type":    "done",
		"payload": map[string]interface{}{"final_result": "completed"},
	})
	fmt.Fprintf(w, "event: task.done\ndata: %s\n\n", done)
	flusher.Flush()
}

func handler(w http.ResponseWriter, r *http.Request) {
	var body map[string]interface{}
	json.NewDecoder(r.Body).Decode(&body)

	if method, _ := body["method"].(string); method == "message/stream" {
		handleStream(w, r)
		return
	}
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]interface{}{
		"jsonrpc": "2.0",
		"id":      body["id"],
		"result":  map[string]interface{}{"echo": body},
	})
}

func registerAgent() {
	card := map[string]interface{}{
		"agent_id":           agentID,
		"name":               "Slow Stream Agent",
		"description":        "Emits events slowly for disconnect testing",
		"version":            "1.0.0",
		"url":                fmt.Sprintf("http://localhost:%d/", agentPort),
		"skills":             []interface{}{},
		"defaultInputModes":  []string{"text"},
		"defaultOutputModes": []string{"text"},
		"protocolVersions":   []string{"1.0"},
	}
	data, _ := json.Marshal(card)
	resp, err := http.Post(kubemqURL+"/agents/register", "application/json", bytes.NewReader(data))
	if err != nil {
		fmt.Fprintf(os.Stderr, "Registration failed: %v\n", err)
		return
	}
	resp.Body.Close()
	fmt.Printf("Registered: %d\n", resp.StatusCode)
}

func main() {
	http.HandleFunc("/", handler)

	ln, err := net.Listen("tcp", fmt.Sprintf(":%d", agentPort))
	if err != nil {
		fmt.Fprintf(os.Stderr, "Listen failed: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("Agent listening on port %d\n", agentPort)

	go http.Serve(ln, nil)

	registerAgent()

	sig := make(chan os.Signal, 1)
	signal.Notify(sig, os.Interrupt)
	<-sig
	fmt.Println("\nShutting down")
}

File: examples/typescript/streaming/client-disconnect/agent.ts
import express from "express";

const KUBEMQ_URL = "http://localhost:9090";
const AGENT_ID = "disconnect-agent-01";
const AGENT_PORT = 18080;

const app = express();
app.use(express.json());

app.post("/", async (req, res) => {
  const accept = req.headers.accept || "";

  if (accept.includes("text/event-stream") || req.body.method === "message/stream") {
    res.setHeader("Content-Type", "text/event-stream");
    res.setHeader("Cache-Control", "no-cache");
    res.setHeader("Connection", "keep-alive");
    res.flushHeaders();

    for (let i = 1; i <= 10; i++) {
      if (res.writableEnded) {
        console.log(`Client disconnected at event ${i}`);
        return;
      }
      const event = {
        type: "status_update",
        payload: { status: "working", progress: i, total: 10 },
      };
      res.write(`event: task.status\ndata: ${JSON.stringify(event)}\n\n`);
      console.log(`Sent event ${i}/10`);
      await new Promise((r) => setTimeout(r, 1000));
    }

    const done = { type: "done", payload: { final_result: "completed" } };
    res.write(`event: task.done\ndata: ${JSON.stringify(done)}\n\n`);
    res.end();
  } else {
    res.json({ jsonrpc: "2.0", id: req.body.id, result: { echo: req.body } });
  }
});

app.listen(AGENT_PORT, async () => {
  console.log(`Disconnect agent listening on port ${AGENT_PORT}`);
  console.log("(emits 10 events slowly — client disconnects after 2)");

  const card = {
    agent_id: AGENT_ID,
    name: "Disconnect Agent",
    description: "Emits events slowly for client disconnect testing",
    version: "1.0.0",
    url: `http://localhost:${AGENT_PORT}/`,
    skills: [],
    defaultInputModes: ["text"],
    defaultOutputModes: ["text"],
    protocolVersions: ["1.0"],
  };

  const resp = await fetch(`${KUBEMQ_URL}/agents/register`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(card),
  });
  const data = await resp.json();
  console.log(`Registered: ${data.agent_id}`);
});
